Alex_RTOS_Sailboat/RTOS_framework_v1/src/sail_gps.c

```diff
@@ -0,0 +1,344 @@
+
+	//DEBUG_Write("WS: %s\r\n", msg_buffer);
+
+	//assign NMEA string prefix to raw_data type
+	if (!get_NMEA_type(&raw_data.type)) {
+		return STATUS_DATA_NOT_NEEDED;
+	}
+
+	// Get each argument after the type
+	uint8_t arg_count = 0;
+
+	while ((msg_ptr = strtok(NULL, ",")) != NULL) {
+		//store msg_ptr as float value into arg array ..
+		//if *msg_ptr is alphabetic, store directly, else convert string to float value
+		raw_data.args[arg_count++] = isalpha(*msg_ptr) ? *msg_ptr : atof(msg_ptr);
+
+		if (arg_count == GPS_MSG_MAX_ARGS) break;
+	}
+
+	/*
+	// Compare the argument count to its expected value
+	if (arg_count != WEATHERSENSOR_arg_counts[raw_data.type]) {
+		return STATUS_ERR_BAD_DATA;
+	}
+	*/
+
+	// Parse the message
+	if (GPS_ExtractMsg(msg, &raw_data) != STATUS_OK) {
+		return STATUS_ERR_BAD_DATA;
+	}
+
+	return STATUS_OK;
+}
+
+
+int last_type = 1337;
+int vals[4];
+
+//extern int was_here;
+static enum status_code GPS_ExtractMsg(NMEA_GenericMsg* msg, GPS_MsgRawData* data) {
+	msg->type = data->type;
+	last_type = data->type;
+
+	// args[0] is the first argument after the NMEA type
+	//ex. GPGGA,<arg[0]>,<arg[1]>...
+	switch (data->type)
+	{
+	case eGPGGA:
+		msg->fields.gpgga.lat.lat = data->args[1];
+		msg->fields.gpgga.lat.ns = ((char)data->args[2] == 'N') ? north : south;
+		msg->fields.gpgga.lon.lon = data->args[3];
+		msg->fields.gpgga.lon.we = ((char)data->args[4] == 'W') ? west : east;
+		msg->fields.gpgga.alt = data->args[8];
+		vals[0] = 1;
+		break;
+
+		/* case eGPVTG:
+			msg->fields.gpvtg.course_over_ground = data->args[0];
+			break; */
+
+	case eWIMWV:
+		msg->fields.wimwv.wind_dir_rel = data->args[0];
+		msg->fields.wimwv.wind_speed_ms = data->args[2];
+		break;
+
+		// This the YXXDR-B type NMEA message
+	case eYXXDR:
+		if (data->args[0] == 'A') {
+			msg->fields.yxxdr.pitch_deg = data->args[1];
+			msg->fields.yxxdr.roll_deg = data->args[5];
+			vals[3] = 1;
+		}
+
+		break;
+
+	case eHCHDT:
+		msg->fields.hchdt.bearing = data->args[0];
+		vals[1] = 1;
+		break;
+		/*
+		case eWIMWD:
+			msg->fields.wimwd.wind_dir_true = data->args[0];
+			msg->fields.wimwd.wind_dir_mag = data->args[2];
+			msg->fields.wimwd.wind_speed_knot = data->args[4];
+			msg->fields.wimwd.wind_speed_ms = data->args[6];
+			vals[2] = 1;
+		break;
+
+		case eYXXDR:
+		//the yxxdr type we are interested in
+			if(data->args[0] == 'A') {
+				msg->fields.yxxdr.pitch_deg = data->args[1];
+				msg->fields.yxxdr.roll_deg = data->args[5];
+				vals[3] = 1;
+			//	DEBUG_Write("Assigning data!\n");
+			}
+
+		break;
+
+		//the remaining message cases are listed below
+
+		case eGPDTM:
+		break;
+		case eGPGLL:
+		break;
+		case eGPGSA:
+		break;
+		case eGPGSV:
+		break;
+		case eGPRMC:
+		break;
+		case eGPVTG:
+		break;
+		case eGPZDA:
+		break;
+		case eHCHDG:
+		break;
+		case eHCTHS:
+		break;
+		case eTIROT:
+		break;
+		case eWIMDA:
+		break;
+		case eWIMWV:
+		break;
+		case eWIMWR:
+		break;
+		case eWIMWT:
+		break;
+		*/
+
+
```

Comment:
Unless we are planning to add all this types lets consider removing this comments. Lets make this code great again ;)